File: go/min_misrate.go
package pragmastat

import "math"

// minAchievableMisrateOneSample computes the minimum achievable misrate
// for one-sample signed-rank based bounds.
// Returns 2^(1-n) which is the smallest possible misrate for a sample of size n.
func minAchievableMisrateOneSample(n int) (float64, error) {
	if n <= 0 {
		return 0, &AssumptionError{Violation: Violation{ID: Domain, Subject: SubjectX}}
	}
	return math.Pow(2, float64(1-n)), nil
}

// minAchievableMisrateTwoSample computes the minimum achievable misrate
// for two-sample Mann-Whitney based bounds.
func minAchievableMisrateTwoSample(n, m int) (float64, error) {
	if n <= 0 {
		return 0, NewDomainError(SubjectX)
	}
	if m <= 0 {
		return 0, NewDomainError(SubjectY)
	}
	var binom float64
	if n+m < maxAcceptableBinomN {
		binom = float64(binomialCoefficient(n+m, n))
	} else {
		binom = binomialCoefficientFloat(float64(n+m), float64(n))
	}
	return 2.0 / binom, nil
}

File: ts/src/minMisrate.ts
/**
 * MinAchievableMisrate functions for bounds validation.
 */

import { AssumptionError } from './assumptions';
import { binomialCoefficient } from './binomial';

/**
 * Computes the minimum achievable misrate for one-sample bounds.
 *
 * For a sample of size n, the minimum achievable misrate is 2^(1-n),
 * which corresponds to the probability of the most extreme configuration
 * in the Wilcoxon signed-rank distribution.
 *
 * @param n Sample size (must be positive)
 * @returns Minimum achievable misrate
 */
export function minAchievableMisrateOneSample(n: number): number {
  if (n <= 0) {
    throw AssumptionError.domain('x');
  }
  return Math.pow(2, 1 - n);
}

/**
 * Computes the minimum achievable misrate for two-sample Mann-Whitney based bounds.
 *
 * @param n Size of first sample (must be positive)
 * @param m Size of second sample (must be positive)
 * @returns Minimum achievable misrate
 */
export function minAchievableMisrateTwoSample(n: number, m: number): number {
  if (n <= 0) {
    throw AssumptionError.domain('x');
  }
  if (m <= 0) {
    throw AssumptionError.domain('y');
  }
  // Shares one binomial with the exact pairwise-margin distribution (see binomial.ts):
  // the floor and the distribution it guards must be computed by the same route.
  return 2.0 / binomialCoefficient(n + m, n);
}
